portal/models/reporting.py

```diff
@@ -0,0 +1,81 @@
+"""Reporting statistics and data module"""
+from collections import defaultdict
+
+from ..dogpile import dogpile_cache
+from .fhir import CC
+from .intervention import Intervention
+from .procedure_codes import known_treatment_started
+from .procedure_codes import known_treatment_not_started
+from .role import ROLE
+from .user import User
+
+
+@dogpile_cache.region('hourly')
+def get_reporting_stats():
+    """Cachable interface for expensive reporting data queries
+
+    The following code is only run on a cache miss.
+
+    """
+    stats = {}
+    stats['roles'] = defaultdict(int)
+    stats['patients'] = defaultdict(int)
+    stats['interventions'] = defaultdict(int)
+    stats['intervention_access'] = defaultdict(int)
+    stats['intervention_reports'] = defaultdict(int)
+    stats['organizations'] = defaultdict(int)
+    stats['registrations'] = []
+    stats['encounters'] = defaultdict(list)
+
+    interventions = Intervention.query.all()
+
+    for user in User.query.filter_by(active=True):
+        if ROLE.TEST in [r.name for r in user.roles]:
+            continue
+
+        for role in user.roles:
+            stats['roles'][role.name] += 1
+            if role.name == 'patient':
+                if not any((obs.codeable_concept == CC.BIOPSY
+                            and obs.value_quantity.value)
+                           for obs in user.observations):
+                    stats['patients']['pre-dx'] += 1
+                elif known_treatment_not_started(user):
+                    stats['patients']['dx-nt'] += 1
+                elif known_treatment_started(user):
+                    stats['patients']['dx-t'] += 1
+                if any((obs.codeable_concept == CC.PCaLocalized
+                        and obs.value_quantity == CC.FALSE_VALUE)
+                       for obs in user.observations):
+                    stats['patients']['meta'] += 1
+
+        for interv in interventions:
+            desc = interv.description
+            if desc == 'Decision Support':
+                desc = 'Decision Support P3P'
```

Comment:
Hmm - there are multiple `decision support` options.  I'm not sure we should be forcing this to P3P here.  Would it be safer to switch on intervention.name?